### PR TITLE
check all nearby omts for lab tunnels

### LINKED ISSUE #45973: Find Lab Tunnels refugee center mission is bugged and uncompletable.
# Describe the bug

The last Refugee Center doctor mission is bugged at start and can't find a suitable target console for the mission. Thus the mission is unfinishable.

# Steps To Reproduce

1. Do all the Refugee Center doctor mission until the last one.
2. Try to accept the last mission.
3. The game will throw up an error saying the mission will fail.

# Expected behavior

The mission should progress normally.

# Screenshots

# Versions and configuration

- OS: Windows
    - OS Version: MINGW/CYGWIN/MSYS2 on unknown Windows version [Windows 8.1 Pro build 9600]
- Game Version: 0.E-8109-g9953469 [64-bit]
- Graphics Version: Tiles
- Game Language: English [en]
- Mods loaded: [
    Dark Days Ahead [dda],
    Disable NPC Needs [no_npc_food],
    C.R.I.T Expansion Mod [crt_expansion],
    Blaze Industries [blazeindustries],
    Alternative Map Key [alt_map_key],
    Graphical Overmap [Graphical_Overmap],
    sees-player icon, retrodays [sees_player_retro],
    sees-player icon, HitButton_iso [sees_player_hitbutton],
    Mythical Martial Arts [MMA],
    More Survivor Stuff [more_survivor_stuff],
    Fuji's Military Profession Pack [fuji_mpp],
    Mutation Changes [Mutation_Changes],
    SpeedyDex [speedydex],
    Stats Through Skills [StatsThroughSkills],
    Blaze Industries - UndeadPeople Tileset [BlazeIndustriesUDP],
    C.R.I.T Expansion Mod - UndeadPeople Tileset [crt_expansionUDP]
]

# Additional context

Error Message:

 DEBUG    : Could not find a computer in the lab train depot, mission will fail.

 FUNCTION : static void mission_start::reveal_lab_train_depot(mission*)
 FILE     : src/mission_start.cpp
 LINE     : 738

[Sabattis.zip](https://github.com/CleverRaven/Cataclysm-DDA/files/5676294/Sabattis.zip)

### PR DESCRIPTION
**Pull request by KorGgenT**
#### Summary
Bugfixes "fix lab tunnels mission not appearing if the console is destroyed"

#### Purpose of change
Fixes #45973

#### Describe the solution
instead of only checking one OMT, check for each one starting with the closest. the closest one with a fixed console will be the mission target.

#### Testing
loaded the save in the attached issue, talked to the doc. the mission target populates properly.

### CODE REVIEW

File: src/mission_start.cpp [hunk 1]
```diff
@@ -735,15 +729,42 @@ void mission_start::reveal_lab_train_depot( mission *miss )
...
+
+    const std::vector<tripoint_abs_omt> all_omts_near = overmap_buffer.find_all( loc, params );
+    // sort it by range
+    std::multimap<int, tripoint_abs_omt> omts_by_range;
+    for( const tripoint_abs_omt &location : all_omts_near ) {
+        omts_by_range.emplace( rl_dist( loc, location ), location );
+    }
+    for( const std::pair<int, tripoint_abs_omt> &location : omts_by_range ) {
+        if( has_console( location.second, mission_id ) ) {
+            place = location.second;
```
Comment (BrettDong) [inline src/mission_start.cpp:765]: No break here?

Comment (KorGgenT) [inline src/mission_start.cpp:765]:
```suggestion
            place = location.second;
            break;
```

File: src/mission_start.cpp [hunk 2]
```diff
@@ -735,15 +729,42 @@ void mission_start::reveal_lab_train_depot( mission *miss )
...
-        return;
+        return false;
     }
 
     computer *tmpcomp = compmap.computer_at( *comppoint );
-    tmpcomp->set_mission( miss->get_id() );
+    tmpcomp->set_mission( mission_id );
     tmpcomp->add_option( _( "Download Routing Software" ), COMPACT_DOWNLOAD_SOFTWARE, 0 );
 
     compmap.save();
```
Comment (KorGgenT) [inline src/mission_start.cpp:739]:
```suggestion
    compmap.save();
    return true;
```

File: src/mission_start.cpp [hunk 3]
```diff
@@ -735,15 +729,44 @@ void mission_start::reveal_lab_train_depot( mission *miss )
...
+    temp_pair.second = ot_match_type::type;
+    params.types.push_back( temp_pair );
+
+    const std::vector<tripoint_abs_omt> all_omts_near = overmap_buffer.find_all( loc, params );
+    // sort it by range
+    std::multimap<int, tripoint_abs_omt> omts_by_range;
+    for( const tripoint_abs_omt &location : all_omts_near ) {
+        omts_by_range.emplace( rl_dist( loc, location ), location );
+    }
+    for( const std::pair<int, tripoint_abs_omt> &location : omts_by_range ) {
```
Comment (KorGgenT) [inline src/mission_start.cpp:764]:
```suggestion
    for( const std::pair<int, const tripoint_abs_omt> &location : omts_by_range ) {
```

Comment (KorGgenT) [reply src/mission_start.cpp:764]: is this what the github error means to do?

Comment (actual-nh) [reply src/mission_start.cpp:764]: I am guessing so - either that or remove the `&` before `location`? (I have trouble enough with pointers/references in C; C++ ones do not make much sense to me...)

File: src/mission_start.cpp [hunk 4]
```diff
@@ -735,15 +729,44 @@ void mission_start::reveal_lab_train_depot( mission *miss )
...
+    temp_pair.second = ot_match_type::type;
+    params.types.push_back( temp_pair );
+
+    const std::vector<tripoint_abs_omt> all_omts_near = overmap_buffer.find_all( loc, params );
+    // sort it by range
+    std::multimap<int, tripoint_abs_omt> omts_by_range;
+    for( const tripoint_abs_omt &location : all_omts_near ) {
+        omts_by_range.emplace( rl_dist( loc, location ), location );
+    }
+    for( const std::pair<int, const tripoint_abs_omt> &location : omts_by_range ) {
```
Comment (BrettDong) [inline src/mission_start.cpp:764]: This is the right way to iterate through a `multimap` in ranged based for loop if not using `auto`
```suggestion
    for( const std::pair<const int, tripoint_abs_omt> &location : omts_by_range ) {
```

File: src/mission_start.cpp [hunk 5]
```diff
@@ -735,15 +729,44 @@ void mission_start::reveal_lab_train_depot( mission *miss )
...
+    loc.z() = -4;  // tunnels are at z = -4
+    tripoint_abs_omt place;
+    const int mission_id = miss->get_id();
+
+    omt_find_params params;
+    std::pair<std::string, ot_match_type> temp_pair;
+    temp_pair.first = "lab_train_depot";
+    temp_pair.second = ot_match_type::type;
+    params.types.push_back( temp_pair );
+
```
Comment (BrettDong) [inline src/mission_start.cpp:757]:
```suggestion
    omt_find_params params = {{ {{ std::make_pair( "lab_train_depot", ot_match_type::type ) }} }};
```

**Review by BrettDong — APPROVED** [review]
Tested and confirm fix.

**Pull request merged**